File: platform/libs/expression-dsl/src/main/java/com/novaforge/expression/Lexer.java
package com.novaforge.expression;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

/** expr/v1 tokenizer (PHASE-2 Annex A): literals, identifiers, operators, punctuation. */
final class Lexer {

    record Token(String kind, Object value, int position) {

        static Token of(String kind, int position) {
            return new Token(kind, null, position);
        }

        static Token of(String kind, Object value, int position) {
            return new Token(kind, value, position);
        }
    }

    private final String source;
    private final List<Token> tokens = new ArrayList<>();
    private int index;

    Lexer(String source) {
        this.source = source == null ? "" : source;
    }

    List<Token> lex() {
        while (index < source.length()) {
            char c = source.charAt(index);
            if (Character.isWhitespace(c)) {
                index++;
                continue;
            }
            int start = index;
            if (Character.isLetter(c) || c == '_') {
                lexIdentifier(start);
            } else if (Character.isDigit(c)) {
                lexNumber(start);
            } else if (c == '\'') {
                lexString(start);
            } else {
                lexOperator(start);
            }
        }
        tokens.add(Token.of("eof", index));
        return tokens;
    }

    private void lexIdentifier(int start) {
        while (index < source.length()
                && (Character.isLetterOrDigit(source.charAt(index)) || source.charAt(index) == '_')) {
            index++;
        }
        tokens.add(Token.of("ident", source.substring(start, index), start));
    }

    private void lexNumber(int start) {
        boolean decimal = false;
        while (index < source.length()) {
            char c = source.charAt(index);
            if (Character.isDigit(c)) {
                index++;
            } else if (c == '.' && index + 1 < source.length()
                    && Character.isDigit(source.charAt(index + 1)) && !decimal) {
                decimal = true;
                index++;
            } else {
                break;
            }
        }
        // Exact decimal literals — arbitrary precision, never binary float (Annex A).
        tokens.add(Token.of("number", new BigDecimal(source.substring(start, index)), start));
    }

    private void lexString(int start) {
        index++;   // opening quote
        StringBuilder value = new StringBuilder();
        while (index < source.length() && source.charAt(index) != '\'') {
            value.append(source.charAt(index));
            index++;
        }
        if (index >= source.length()) {
            throw new ExpressionException("unterminated string literal at position " + start);
        }
        index++;   // closing quote
        tokens.add(Token.of("string", value.toString(), start));
    }

    private void lexOperator(int start) {
        char c = source.charAt(index);
        String two = index + 1 < source.length()
                ? source.substring(index, index + 2) : null;
        if (two != null && (two.equals("==") || two.equals("!=") || two.equals("<=")
                || two.equals(">=") || two.equals("&&") || two.equals("||"))) {
            tokens.add(Token.of(two, start));
            index += 2;
            return;
        }
        switch (c) {
            case '<', '>', '+', '-', '*', '/', '!', '(', ')', ',', '.' -> {
                tokens.add(Token.of(String.valueOf(c), start));
                index++;
            }
            default -> throw new ExpressionException(
                    "unexpected character '" + c + "' at position " + start);
        }
    }
}

File: frontend/shared/src/expression/lexer.ts
import { Decimal } from "./decimal.ts";
import { ExpressionError } from "./values.ts";

/**
 * expr/v1 tokenizer — the TS twin of the JVM Lexer (PHASE-2 Annex A): literals,
 * identifiers, operators, punctuation. `${…}` and bitwise operators are rejected
 * here (template interpolation is host-side; the scalar grammar has no bitwise ops).
 */

export interface Token {
    kind: string;
    value?: string | Decimal;
    position: number;
}

const TWO_CHAR = new Set(["==", "!=", "<=", ">=", "&&", "||"]);
const SINGLE = new Set(["<", ">", "+", "-", "*", "/", "!", "(", ")", ",", "."]);

export function lex(source: string): Token[] {
    const tokens: Token[] = [];
    let index = 0;
    const at = (i: number) => source.charAt(i);
    while (index < source.length) {
        const c = at(index);
        if (/\s/.test(c)) {
            index++;
            continue;
        }
        const start = index;
        if (/[A-Za-z_]/.test(c)) {
            index++;
            while (index < source.length && /[A-Za-z0-9_]/.test(at(index))) index++;
            tokens.push({ kind: "ident", value: source.slice(start, index), position: start });
            continue;
        }
        if (/[0-9]/.test(c)) {
            let decimal = false;
            index++;
            while (index < source.length) {
                const d = at(index);
                if (/[0-9]/.test(d)) {
                    index++;
                } else if (d === "." && !decimal && /[0-9]/.test(at(index + 1))) {
                    decimal = true;
                    index++;
                } else {
                    break;
                }
            }
            // Exact decimal literals — arbitrary precision, never binary float.
            tokens.push({
                kind: "number",
                value: Decimal.parse(source.slice(start, index)),
                position: start,
            });
            continue;
        }
        if (c === "'") {
            index++;
            let value = "";
            while (index < source.length && at(index) !== "'") {
                value += at(index);
                index++;
            }
            if (index >= source.length) {
                throw new ExpressionError(`unterminated string literal at position ${start}`);
            }
            index++;
            tokens.push({ kind: "string", value, position: start });
            continue;
        }
        const two = source.slice(index, index + 2);
        if (TWO_CHAR.has(two)) {
            tokens.push({ kind: two, position: start });
            index += 2;
            continue;
        }
        if (SINGLE.has(c)) {
            tokens.push({ kind: c, position: start });
            index++;
            continue;
        }
        throw new ExpressionError(`unexpected character '${c}' at position ${start}`);
    }
    tokens.push({ kind: "eof", position: index });
    return tokens;
}
